### PR TITLE
support protocol 2.1 and 3.0

### PR DESCRIPTION
**Pull request by danrubel**
@devoncarew 
+ support protocol 2.1 and 3.0
+ copy addBreakpoint into VmServiceBase in preparation for moving to protocol 3.0

### CODE REVIEW

File: java/src/org/dartlang/vm/service/VmServiceBase.java [hunk 1]
```diff
@@ -190,6 +193,18 @@ public static VmService localConnect(int port) throws IOException {
   RequestSink requestSink;
 
   /**
+   * The [addBreakpoint] RPC is used to add a breakpoint at a specific line of some script. This
+   * call works for protocol versions 2.1 and 3.0
```
Comment (devoncarew) [inline java/src/org/dartlang/vm/service/VmServiceBase.java]: The protocol went from 2.0 to 3.0 - there wasn't a 2.1.

File: java/src/org/dartlang/vm/service/VmServiceBase.java [hunk 2]
```diff
@@ -130,17 +131,19 @@ public void onError(RPCError error) {
...
-          Logging.getLogger().logError(msg);
-          errMsg[0] = msg;
-        } else if (response.getMinor() != VmService.versionMinor) {
-          Logging.getLogger().logInformation(
-              "Minor difference in protocol version: client=" + VmService.versionMajor + "."
-                  + VmService.versionMinor + " vm=" + response.getMajor() + "."
-                  + response.getMinor());
+        int major = response.getMajor();
+        int minor = response.getMinor();
+        if (major != VmService.versionMajor || minor != VmService.versionMinor) {
```
Comment (devoncarew) [inline java/src/org/dartlang/vm/service/VmServiceBase.java]: We should probably ignore the minor version when checking for compatibility.

File: java/src/org/dartlang/vm/service/VmServiceBase.java [hunk 3]
```diff
@@ -130,17 +131,19 @@ public void onError(RPCError error) {
...
+          if ((major == 2 && minor == 1) || (major == 3 && minor == 0)) {
```
Comment (devoncarew) [inline java/src/org/dartlang/vm/service/VmServiceBase.java]: Is the VM reporting `2.1`? I don't see it in the spec doc.

Comment (danrubel) [inline java/src/org/dartlang/vm/service/VmServiceBase.java]: That's what Dart SDK VM 1.12.x stable is reporting.

File: java/src/org/dartlang/vm/service/VmServiceBase.java (hunk 1 above)
Comment (danrubel) [inline java/src/org/dartlang/vm/service/VmServiceBase.java]: The spec is the "theory" and the VM is the "reality". In theory, theory and reality are the same.... but the Dart SDK 1.12 stable VM is reporting protocol version 2.1.

File: java/src/org/dartlang/vm/service/VmServiceBase.java (hunk 2 above)
Comment (danrubel) [inline java/src/org/dartlang/vm/service/VmServiceBase.java]: Fair enough. Done.

**Comment by danrubel** [comment]
@devoncarew Updated. PTAL.

File: java/src/org/dartlang/vm/service/VmServiceBase.java (hunk 1 above)
Comment (devoncarew) [inline java/src/org/dartlang/vm/service/VmServiceBase.java]: @turnidge @johnmccutchan @rmacnak-google, fyi, it looks like the spec version in the markdown file and the one reported by the VM has been out of sync at times.

**Comment by devoncarew** [comment]
lgtm

**Pull request merged**